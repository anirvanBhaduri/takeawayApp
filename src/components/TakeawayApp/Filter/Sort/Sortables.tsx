import * as React from 'react';

import {RestaurantProps} from "../../RestaurantList/Restaurant/Restaurant";

/**
 * Defines the content of sortables.
 *
 * Each sortable has a title and a filter method.
 *
 * @interface SortFilterProps
 */
export interface SortFilterProps {
    title: string;
    filter: (
        firstList: RestaurantProps,
        secondList: RestaurantProps,
        sortKey: string
    ) => number;
    sortValue: (props: RestaurantProps, sortKey: string) => number;
}

/**
 * A generic sort filter method.
 *
 * Used for most sorts. For this filter function,
 * we are going with the assumption that the HIGHER number should be
 * closer to the top and the LOWER should be close to the bottom.
 *
 * @param firstList
 * @param secondList
 * @param sortKey
 *
 * @return number
 */
const sortFilter = (
    firstList: RestaurantProps,
    secondList: RestaurantProps,
    sortKey: string
): number => {
    return secondList.sortingValues[sortKey] - firstList.sortingValues[sortKey];
};

/**
 * Filter method for sorting by top restaurants.
 *
 * @param firstList
 * @param secondList
 *
 * @return number
 */
const topRestaurant = (
    firstList: RestaurantProps,
    secondList: RestaurantProps
): number => {
    // extract values required for formula - firstList
    const {
        distance: fDistance,
        popularity: fPopularity,
        ratingAverage: fRatingAverage
    } = firstList.sortingValues;

    // extract values required for formula - secondList
    const {
        distance: sDistance,
        popularity: sPopularity,
        ratingAverage: sRatingAverage
    } = secondList.sortingValues;

    const fTotal = (fDistance * fPopularity) + fRatingAverage;
    const sTotal = (sDistance * sPopularity) + sRatingAverage;
    return sTotal - fTotal;
};

/**
 * Generic sortValue function.
 *
 * Gets the sortValue based on a sortKey and a given RestaurantProp.
 *
 * @param props
 * @param sortKey
 *
 * @return number
 */
const sortValue = (props: RestaurantProps, sortKey: string): number => {
    return props.sortingValues[sortKey];
};

/**
 * Sort types.
 *
 * This object defines the title and filter method for the sort
 * values.
 */
export const sortables: {[index: string]: SortFilterProps} = {
    bestMatch: {
        title: 'Best Match',
        filter: sortFilter,
        sortValue: sortValue,
    },
    newest: {
        title: 'Newest',
        filter: sortFilter,
        sortValue: sortValue,
    },
    ratingAverage: {
        title: 'Rating Average',
        filter: sortFilter,
        sortValue: sortValue,
    },
    distance: {
        title: 'Distance',
        filter: sortFilter,
        sortValue: sortValue,
    },
    popularity: {
        title: 'Popularity',
        filter: sortFilter,
        sortValue: sortValue,
    },
    averageProductPrice: {
        title: 'Average Product Price',
        filter: sortFilter,
        sortValue: sortValue,
    },
    deliveryCosts: {
        title: 'Delivery Costs',
        filter: sortFilter,
        sortValue: sortValue,
    },
    minCost: {
        title: 'Min Cost',
        filter: sortFilter,
        sortValue: sortValue,
    },
    topRestaurant: {
        title: 'Top Restaurant',
        filter: topRestaurant,
        sortValue: (props: RestaurantProps): number => {
            const {
                distance,
                popularity,
                ratingAverage
            } = props.sortingValues;

            return (distance * popularity) + ratingAverage;
        }
    }
};
