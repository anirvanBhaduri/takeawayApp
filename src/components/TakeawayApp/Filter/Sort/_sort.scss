/// Sort component styles.
///
/// @author Anirvan Bhaduri
/// @since 10th Dec 2018

@import '../../UI/colours';
@import '../../UI/dimensions';
@import '../../UI/mixins';

// The sort dropdown's styles
.sort {
  @include centre-container();
  margin-top: 10px;
  margin-bottom: 20px;
}

// The sort select's styles
.sort__select {
  @include nice-border();
  color: white;
  background: $takeaway-color;
  width: 100%;
  font-size: inherit;
}

// The sort dropdown option's styles
.sort__option {
  @include nice-border();
  color: white;
  background: $takeaway-color;
  font-size: inherit;
}
