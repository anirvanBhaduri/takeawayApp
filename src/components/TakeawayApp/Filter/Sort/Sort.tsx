import * as React from 'react';

import {Filter, FilterProps} from "../Filter";
import {sortables} from "./Sortables";
import {RestaurantProps} from "../../RestaurantList/Restaurant/Restaurant";

/**
 * Defines props required by Sort.
 *
 * We need sortType to determine which
 * filter method to run.
 *
 * @interface SortProps
 */
interface SortProps extends FilterProps {
    sortType: string,
}

/**
 * Render the sort component.
 *
 * @return JSX
 */
export class Sort extends Filter<SortProps> {

    /**
     * A list of sortables.
     *
     * @property sortables
     */
    sortables: JSX.Element[];

    /**
     * The filter type.
     *
     * Determines which filter method to run for sorting.
     *
     * @property sortType
     */
    sortType: string;

    /**
     * Sort constructor.
     *
     * Lets populate sortables here.
     *
     * @param props
     */
    constructor(props: any) {
        super(props);

        this.sortType = this.props.sortType;
        this.sortables = Object.keys(sortables).map((key) => {
            return (
                <option
                    key={key}
                    value={key}
                >
                    {sortables[key].title}
                </option>
            );
        });
    }

    /**
     * Set the sort type.
     *
     * @param sortType
     */
    setSortType(sortType: string) {
        this.sortType = sortType;
    }

    /**
     * Apply the sort filter.
     *
     * @param list
     *
     * @return RestaurantProps
     */
    applyFilter(list: RestaurantProps[]): RestaurantProps[] {
        const sortType = this.sortType;

        list.sort((first: RestaurantProps, second: RestaurantProps): number => {
            return sortables[sortType].filter(first, second, sortType);
        });

        return list;
    }

    /**
     * Render the Sort dropdown.
     *
     * @return JSX
     */
    render() {
        const handler = this.props.filterHandler;

        return (
            <select onChange={handler}>
                {this.sortables}
            </select>
        );
    }
}