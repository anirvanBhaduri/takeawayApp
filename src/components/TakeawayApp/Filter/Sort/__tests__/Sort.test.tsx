import * as React from 'react';
import {shallow} from "enzyme";

// SUT
import {Sort} from "../Sort";

/**
 * Expected data that is pre sorted. Output from
 * test should result in this expected data.
 */
export const sortedData = [
    {
        favourite: false,
        name: 'what',
        status: 'open',
    }
];

/**
 * Sort component tests.
 *
 * @covers Sort
 */
describe('Sort Component Tests', () => {

    // /**
    //  * Apply the Sort filter.
    //  *
    //  * @covers Sort.applyFilter
    //  */
    // it('should apply the search filter to the restaurant props array', () => {
    //     const search = new Sort({
    //         searchTerm: 'wh',
    //         filterHandler: () => {}
    //     });
    //
    //     const sorted = search.applyFilter([
    //         {
    //             favourite: false,
    //             name: 'what',
    //             status: 'open',
    //             clickHandler: () => {},
    //         },
    //         {
    //             favourite: true,
    //             name: 'test',
    //             status: 'closed',
    //             clickHandler: () => {},
    //         }
    //     ]);
    //
    //     // Deep copy check using convoluted cloning logic.
    //     // I think we need to do this because the objects themselves have interesting
    //     // props being added to them somehow... perhaps a side effect of
    //     // using typescript?
    //     expect(JSON.parse(JSON.stringify(sorted))).toEqual(sortedData);
    // });

    /**
     * Sort renders correctly.
     *
     * @covers Sort.render
     */
    it('should render correct snapshot of Sort Component', () => {
        const wrapper = shallow(
            <Sort sortType={'bestMatch'} filterHandler={() => {}}/>
        );

        expect(wrapper).toMatchSnapshot();
    });
});
