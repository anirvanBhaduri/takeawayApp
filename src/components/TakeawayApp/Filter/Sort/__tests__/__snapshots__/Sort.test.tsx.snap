// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort Component Tests should render correct snapshot of Sort Component 1`] = `
<div
  className="sort"
>
  <select
    className="sort__select"
    onChange={[Function]}
  >
    <option
      className="sort__option"
      key="bestMatch"
      value="bestMatch"
    >
      Best Match
    </option>
    <option
      className="sort__option"
      key="newest"
      value="newest"
    >
      Newest
    </option>
    <option
      className="sort__option"
      key="ratingAverage"
      value="ratingAverage"
    >
      Rating Average
    </option>
    <option
      className="sort__option"
      key="distance"
      value="distance"
    >
      Distance
    </option>
    <option
      className="sort__option"
      key="popularity"
      value="popularity"
    >
      Popularity
    </option>
    <option
      className="sort__option"
      key="averageProductPrice"
      value="averageProductPrice"
    >
      Average Product Price
    </option>
    <option
      className="sort__option"
      key="deliveryCosts"
      value="deliveryCosts"
    >
      Delivery Costs
    </option>
    <option
      className="sort__option"
      key="minCost"
      value="minCost"
    >
      Min Cost
    </option>
    <option
      className="sort__option"
      key="topRestaurant"
      value="topRestaurant"
    >
      Top Restaurant
    </option>
  </select>
</div>
`;
